builtin/audit/http/backend.go

```diff
@@ -0,0 +1,314 @@
+	defer b.saltMutex.Unlock()
+
+	s = b.salt.Load().(*salt.Salt)
+	if s != nil {
+		return s, nil
+	}
+
+	newSalt, err := salt.NewSalt(ctx, b.saltView, b.saltConfig)
+	if err != nil {
+		b.salt.Store((*salt.Salt)(nil))
+		return nil, err
+	}
+
+	b.salt.Store(newSalt)
+	return newSalt, nil
+}
+
+func (b *Backend) GetHash(ctx context.Context, data string) (string, error) {
+	salt, err := b.Salt(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	return audit.HashString(salt, data), nil
+}
+
+func (b *Backend) LogRequest(ctx context.Context, in *logical.LogInput) error {
+	buf := bytes.NewBuffer(make([]byte, 0, 2000))
+	err := b.formatter.FormatRequest(ctx, buf, b.formatConfig, in)
+	if err != nil {
+		return err
+	}
+
+	return b.log(ctx, buf)
+}
+
+func (b *Backend) log(ctx context.Context, buf *bytes.Buffer) error {
+	reader := bytes.NewReader(buf.Bytes())
+
+	client, err := b.getClient()
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, b.uri, reader)
+	if err != nil {
+		return fmt.Errorf("failed to create request: %w", err)
+	}
+
+	req.Header = b.headers.Clone()
+
+	if value := b.headers.Get("User-Agent"); value != "" {
+		req.Header.Add("User-Agent", fmt.Sprintf("OpenBaoAuditor/%s", version.GetVersion().VersionNumber()))
+	}
+
+	resp, err := client.Do(req)
+	if err != nil || resp == nil {
+		return fmt.Errorf("failed to perform request: %w", err)
+	}
+
+	if err := resp.Body.Close(); err != nil {
+		return fmt.Errorf("failed to close body: %w", err)
+	}
+
+	// Redirects and errors will not be accepted.
+	if resp.StatusCode >= 300 {
+		return fmt.Errorf("failed to perform request: status code not 2xx: %v", resp.StatusCode)
+	}
+
+	return nil
+}
+
+func (b *Backend) LogResponse(ctx context.Context, in *logical.LogInput) error {
+	buf := bytes.NewBuffer(make([]byte, 0, 6000))
+	err := b.formatter.FormatResponse(ctx, buf, b.formatConfig, in)
+	if err != nil {
+		return err
+	}
+
+	return b.log(ctx, buf)
+}
+
+func (b *Backend) LogTestMessage(ctx context.Context, in *logical.LogInput, config map[string]string) error {
+	var buf bytes.Buffer
+	temporaryFormatter := audit.NewTemporaryFormatter(config["format"], config["prefix"])
+	if err := temporaryFormatter.FormatRequest(ctx, &buf, b.formatConfig, in); err != nil {
+		return err
+	}
+
+	return b.log(ctx, &buf)
+}
+
+func (b *Backend) getClient() (*http.Client, error) {
+	b.clientLock.RLock()
+	client := b.client
+	b.clientLock.RUnlock()
+
+	if client != nil {
+		return client, nil
+	}
+
+	b.clientLock.Lock()
+	defer b.clientLock.Unlock()
+
+	if b.client != nil {
+		return client, nil
+	}
+
+	// TODO (ascheel): use a different HTTP client.
+	b.client = http.DefaultClient
+
+	return b.client, nil
+}
+
+func (b *Backend) Reload(_ context.Context) error {
+	b.clientLock.Lock()
+	defer b.clientLock.Unlock()
+
+	b.client = nil
+
+	_, err := b.getClient()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (b *Backend) Invalidate(_ context.Context) {
```

Comment:
@wslabosz-reply Was there more you were looking for in invalidation? This is the same as other backends I believe.